atlite/cutout.py

```diff
@@ -1,120 +1,196 @@
+# functionality works fine, even when the pip wheels have been used, only for
+# resampling the sarah dataset it is important to use conda.
+# Refer to
+# https://github.com/pydata/xarray/issues/2535,
+# https://github.com/rasterio/rasterio-wheels/issues/12
+import netCDF4
 
 import xarray as xr
 import numpy as np
 import os, sys
-from six import string_types
+from warnings import warn
+from shapely.geometry import box
+from pathlib import Path
 
 import logging
 logger = logging.getLogger(__name__)
 
-from . import config, datasets
+from . import datasets, utils
 
 from .convert import (convert_and_aggregate, heat_demand, hydro, temperature,
                       wind, pv, runoff, solar_thermal, soil_temperature)
-from .preparation import (cutout_do_task, cutout_prepare,
-                          cutout_produce_specific_dataseries,
-                          cutout_get_meta, cutout_get_meta_view)
-from .gis import compute_indicatormatrix
-
-class Cutout(object):
-    def __init__(self, name=None, cutout_dir=config.cutout_dir, **cutoutparams):
-        self.name = name
+from .gis import GridCells
+from .data import requires_coords, requires_windowed, cutout_prepare
+
+class Cutout:
+
+    dataset_module = None
+
+    def __init__(self, path="unnamed.nc", data=None, cutout_dir=None, **cutoutparams):
+        """
+        Provide an Atlite cutout object.
+
+        Create a cutout object to use atlite operations on it.
+        Based on the provided parameters, atlite first checks
+        whether this cutout already exists on disk and if yes,
+        loads this cutout.
+
+        If the cutout does not yet exist on disk, then atlite
+        creates an "unprepared" cutout object containing all the
+        necessary information for creating the requested cutout,
+        but does not yet create ("prepare") it fully.
+        The process of preparing has to be manually started by
+        `cutout.prepare()`.
+
+        Parameters
+        ----------
+        path : str | path-like
+            NetCDF from which to load or where to store the cutout
+        data : (opt.) xr.Dataset
+            Specify the NetCDF data directly
+        module : ["era5","ncep","cordex","sarah"]
+            The dataset which works as a basis for the cutout
+            creation.
+        time : str | slice
+            Time range to include in the cutout, e.g. "2011" or
+            ("2011-01-05", "2011-01-25")
+        bounds : (opt.) GeoSeries.bounds | DataFrame
+            The outer bounds of the cutout or as a DataFrame
+            containing (min.long, min.lat, max.long, max.lat).
+        x : (opt.) slice
+            Outer longitudinal bounds for the cutout (west, east).
+        y : (opt.) slice
+            Outer latitudinal bounds for the cutout (south, north).
+        time : (opt.) slice
+
+        """
+        if cutout_dir or name:
+            warn("The arguments `cutout_dir` and `name` have been deprecated in 
```

Comment:
Here is a " missing a the end of the line, I'm pushing a small fixup

Reply:
And I also see that it should be a line before
``` 
name = cutoutparams.get("name", None)
```
as it's not defined otherwise

Reply:
No. Also my change and comment were wrong and corona's code was correct.

`name` is currently (`master` branch) the first argument in the `cutout` constructor. in `v0.2` it is superseded by `path`. It has never been in the `cutoutparams` (and should never be).

I'll bring everything together.

This PR is bloating, we should wrap it up and believe in our previous changes ;-)

Reply:
I agree, it's getting really messy. But keep in mind the 'division' by path does not work if it is None.

Reply:
Could you check if this version is now working for you?

Reply:
Yes, works fine! It's also tested in the test script `test_cututout_preparation.py` in #71